macos: increase child process stdio buffer size

verycosy opened the pull request:
## TL;DR
- In libuv, the child process's stdio buffer size is 64KB on Linux, but effectively only 8KB on macOS.
- When large amounts of data are transferred through the child process’s stdio, macOS exhibits significantly slower performance compared to Linux.
- This PR increases the child process’s stdio buffer size from 8KB to 64KB on macOS.

This seems related to the previous discussion in [#3429](https://github.com/nodejs/node/issues/3429).

## Issue
**Environment**
- Apple M4 Pro, 24GB RAM, macOS Sequoia 15.1 (xnu-11215.41.3~3)
- Node.js v22.12.0
- Python v3.12.8
- [reproduced repository](https://github.com/verycosy/repro-libuv-macos-perf)

While processing videos using Node.js and FFmpeg, I noticed significant performance differences between running the same task as a shell command versus implementing it in Node.js.

The following script converts an MP4 video to rawvideo format and then converts it back to MP4 before saving it as a file:

```bash
# script.sh
ffmpeg -i 30fps.mp4 -f rawvideo -pix_fmt rgb24 pipe:1 |
ffmpeg -f rawvideo -pix_fmt rgb24 -s 1920x1080 -r 30 -i pipe:0 -f mp4 -pix_fmt yuv420p -c:v libx264 -movflags empty_moov shell.mp4
```
This operation takes approximately **38 seconds.**
When implemented in Node.js, it looks like this:

```js
// ffmpeg.js
const { spawn } = require('node:child_process');

console.time('conversion');

const srcArgs = ['-i', '30fps.mp4',
                 '-f', 'rawvideo', '-pix_fmt', 'rgb24', 'pipe:1'];
const src = spawn('ffmpeg', srcArgs);

const destArgs = ['-f', 'rawvideo', '-pix_fmt', 'rgb24', '-s', '1920x1080', '-r', '30', '-i', 'pipe:0',
                  '-f', 'mp4', '-pix_fmt', 'yuv420p', '-c:v libx264', '-movflags', 'empty_moov', 'node.mp4'];
const dest = spawn('ffmpeg', destArgs);

src.stdout.pipe(dest.stdin);

src.on('close', () => {
  dest.stdin.end();
});

process.on('exit', () => {
  console.timeEnd('conversion');
});
```
Running this code takes **72 seconds**, nearly **twice as long** as the shell command.
Even when running the same Node.js script inside a Docker container on macOS, it still takes **55 seconds** (although the FFmpeg binaries may differ).

## Experiment
This slow performance seemed unusual, so I investigated the data chunk sizes being transferred through the child process’s stdio.
I found that in macOS, the **chunk size of stdout from the child process is limited to 8192 bytes (8KB)**, whereas in Linux, it is **65536 bytes (64KB).**

```js
// ffmpeg.js
console.log(src.stdout.readableHighWaterMark); // 65536

src.stdout.on('data', (data) => {
  console.log(data.length); // 8192 on macOS, 65536 on Linux
});
```
I also checked whether the same limitation applied to stdin:
```js
// parent.js
const { spawn } = require('node:child_process');

const child = spawn('node', ['child.js']);

child.stdout.on('data', (data) => {
  console.log(data.toString()); // 8192 on macOS, 65536 on Linux
});

child.on('spawn', () => {
  child.stdin.write(Buffer.from('a'.repeat(65536)));
});

// child.js
process.stdin.on('data', (data) => {
  process.stdout.write(Buffer.from(String(data.length)));
});
```
The same limitation was observed in stdin.
This confirms that in macOS, **the maximum chunk size for data transferred through the child process’s stdio is 8KB.**

To further verify, I implemented the same task in Python, which took **40 seconds**—a performance comparable to running the shell command.

```python
# ffmpeg.py
import subprocess
import time

start_time = time.time()

src_args = ["ffmpeg", '-i', '30fps.mp4',
            '-f', 'rawvideo', '-pix_fmt', 'rgb24', 'pipe:1']
src = subprocess.Popen(src_args, stdout=subprocess.PIPE)

dest_args = ["ffmpeg",'-f', 'rawvideo', '-pix_fmt', 'rgb24', '-s', '1920x1080', '-r', '30', '-i', 'pipe:0',
             '-f', 'mp4', '-pix_fmt', 'yuv420p', '-c:v', 'libx264', '-movflags', 'empty_moov', 'python.mp4']
dest = subprocess.Popen(dest_args, stdin=subprocess.PIPE)

chunk_size = 65536

while True:
    data = src.stdout.read(chunk_size)

    if not data:
        break

    dest.stdin.write(data)

dest.stdin.close()
src.wait()
dest.wait()

end_time = time.time()
print(f"conversion time: {end_time - start_time:.2f} seconds")
```

## Solution

Examining the **XNU kernel source code**, I confirmed that the default **AF_UNIX socket buffer size in macOS is 8KB** ([Reference 1](https://github.com/apple-oss-distributions/xnu/blob/rel/xnu-11215/bsd/kern/uipc_usrreq.c#L950-L963), [Reference 2](https://github.com/apple-oss-distributions/xnu/blob/rel/xnu-11215/bsd/kern/uipc_proto.c#L84-L92)).
To address this, I added code to **increase the buffer size to 64KB on macOS**, aligning it with Linux.

As a result, the performance improved significantly.

<img width="717" alt="스크린샷 2025-02-08 오후 11 15 36" src="https://github.com/user-attachments/assets/b57a6403-c107-4769-9795-4861d5bdffd7" />

vtjnash commented:
Why is this a socketpair at all, except when specifically requesting ipc for dealing with nodejs legacy compatibility? It should probably be a uv_pipe (for consistency with libuv Windows), or instead documented as only be a legacy option to uv_spawn (instructing the user to use uv_pipe explicitly)

verycosy commented on src/unix/process.c:
```diff
@@ -199,8 +200,27 @@ static int uv__process_init_stdio(uv_stdio_container_t* container, int fds[2]) {
+    else {
+      ret = uv_socketpair(SOCK_STREAM, 0, fds, 0, 0);
+
+#ifdef __APPLE__
```
I agree that maintaining consistency across all platforms, including Windows, is a good approach. I will investigate how similar buffer settings are handled on Linux and Windows and update the PR accordingly.

Would you suggest applying the same 64KB buffer size setting for uv_pipe on all platforms to ensure consistency?

verycosy commented:
> Why is this a socketpair at all, except when specifically requesting ipc for dealing with nodejs legacy compatibility? It should probably be a uv_pipe (for consistency with libuv Windows), or instead documented as only be a legacy option to uv_spawn (instructing the user to use uv_pipe explicitly)

That makes sense. Would it be acceptable to set the buffer size to 64KB within uv_pipe itself for consistency across platforms?

I’m not yet fully familiar with the entire libuv structure, so I’d appreciate any guidance on whether uv_pipe is the best place to handle this change. Thank you for your patience!

vtjnash commented:
uv_pipe is a proper actual pipe, as the shell and most other programs use, so it may have a different default already

bnoordhuis commented:
> Why is this a socketpair at all [..] It should probably be a uv_pipe

UNIX sockets are bidirectional, pipes are not. It's a thing that comes up sometimes: mildly buggy programs that try to read from stdout or write to stdin, and that works when they refer to a tty... or a UNIX socket.

bnoordhuis commented on include/uv/unix.h:
```diff
@@ -509,4 +509,6 @@ typedef struct {
 #define UV_FS_O_SEQUENTIAL    0
 #define UV_FS_O_TEMPORARY     0
 
+#define UV_SPAWN_BUFFER_SIZE (64 * 1024)
```
Don't define it in uv.h, that's a public header. I also don't think it really needs a define, just use the number directly.

verycosy commented:
Hello, 

Based on your feedback, I did some further investigation.  
As you suggested, I tried using `pipe` instead of `socketpair`, and I found that the buffer size was already 64KB.  
I also agree that using `pipe` is more appropriate for stdio since it is unidirectional.

However, replacing `uv_socketpair` with `uv_pipe` caused multiple tests to fail.  
I suspect this is because the existing implementation relies on `socketpair` for IPC.

There has been previous discussion about whether stdio should use `pipe` instead of `socket`, as seen in this issue:  
→ [https://github.com/nodejs/node/issues/21941](https://github.com/nodejs/node/issues/21941)

Therefore, I believe the core issue is that `spawn` currently uses a socket instead of a pipe for stdio.  
Would it make sense to close this Pull Request and submit a new one that modifies stdio to use `pipe` instead of `socket`?

I’d appreciate your thoughts on this. Thank you!

bnoordhuis commented:
> Would it make sense to close this Pull Request and submit a new one that modifies stdio to use pipe instead of socket?

We can run the libuv <-> node.js integration CI to check if it affects node directly but it won't catch ecosystem fallout further downstream, of which I expect there will be plenty. Node gets used a lot for shell scripting, task runners, etc.

There's a cost/benefit trade-off here that in my opinion leans towards keeping the status quo.

verycosy commented:
I understand! In that case, would it be acceptable to keep `socketpair()` as is, but only increase the buffer size to 64KB as I originally proposed?  

As I mentioned earlier, the performance difference is quite significant, so I believe this change is necessary.  

Would it be okay if I update this PR by addressing the feedback on code style and other minor issues while keeping the buffer size adjustment?

bnoordhuis commented:
Yes, please go ahead.

saghul commented on src/unix/process.c:
```diff
@@ -199,8 +202,24 @@ static int uv__process_init_stdio(uv_stdio_container_t* container, int fds[2]) {
+    if (ret == 0) {
+      for (int i = 0; i < 2; i++) {
+        for (int j = 0; j < 2; j++) {
+          ret = setsockopt(fds[i], SOL_SOCKET, options[j], &size, sizeof(size));
```
I'd ignore the errors here. If we fail to increase the buffer size (not sure why that would happen TBH) so be it

verycosy replied:
```C
#ifdef __APPLE__
    if (ret == 0)
      for (int i = 0; i < 2; i++)
        for (int j = 0; j < 2; j++)
          setsockopt(fds[i], SOL_SOCKET, options[j], &size, sizeof(size));
#endif
```
Would it be acceptable to make this change?

verycosy replied:
I plan to modify the code as follows.

```C
ret = uv_socketpair(SOCK_STREAM, 0, fds, 0, 0);

if (ret == 0)
  for (i = 0; i < 2; i++) {
    setsockopt(fds[i], SOL_SOCKET, SO_RCVBUF, &size, sizeof(size));
    setsockopt(fds[i], SOL_SOCKET, SO_SNDBUF, &size, sizeof(size));
  }

return ret;
```

Additionally, I noticed that the test fails on Unix if the Apple guard is not applied.  
On Unix, the buffer size is automatically set to twice the specified value.  
So, if we set it to `64 * 1024`, it becomes 128KB on Unix.

Wouldn't it be better to apply the Apple guard in this case?

verycosy replied:
Since we have decided to ignore failures when increasing the buffer size,  
should I remove the buffer size verification test that I added?

saghul replied:
Humm, I suspect we don't want to enlarge the size of its already 64k indeed.

verycosy commented:
I have completed all the requested changes.
Could you please take a look?

vtjnash commented:
> There's a cost/benefit trade-off here that in my opinion leans towards keeping the status quo.

Sure, that is why I mainly suggested that we document that this flag is deprecated as a legacy behavior prior to the introduction of `uv_pipe`. That API allows libuv users to avoid various bugs in the way the libuv library implements the compatibility behavior of this (such as this one, https://github.com/nodejs/node/issues/21941, and others)
